### PR TITLE
Add is_default_price_list and discount to master list

### LINKED ISSUE #4703: Use different default sell prices for patient dispensing vs outbound shipments
## Is your feature request related to a problem? Please describe 👀

A requested use case is for one default sell price to be used for when shipping stock to downstream facilities and a different default sell price to be used when dispensing to patients. 

Some Stores may do both - e.g. a hospital sending stock to smaller pharmacies and also dispensing directly to patients. 

## Describe the solution you'd like 🎁

Here's the pricing workflow:

image

## Configuration
( All are items are changes to make in mSupply )
* Create a master list for all items, check the `is_default_price_list` option
* This `is_default_price_list` applies to all stores, and doesn't have to be assigned to them...
* Define prices for each item 
* Create a master list for all discounted items, set the `discount` percentage value
* This discount master list need to be assigned to any stores that should receive the discount rate.
 
## Changes required

* update master lists
  * add `is_default_price_list` to `master_list`
  * add `discount` to `master_list` 
  * and update sync and translation layers
  * add `price` to `master_list_line` and update sync & translation layers

* When stock is issued apply the `is_default_price_list` prices to the invoice line (as appropriate)
* Apply discount to outgoing invoices (apart from dispensing)

**Stocktake**
_No changes_

...OUT OF SCOPE -  POSSIBLE FUTURE REQUIREMENTS...
```
- [ ] If new stock line is introduced by stocktake, it's sell price should be set to price from `is_default_price_list`
- [ ] Stocklines are introduced from either Add Item, or Edit Item (if the stock take is created from a master list) (FRONTEND)
- [ ] Cost prices are assumed to be discounted if discount would apply to this store
- [ ] Sell Prices are not editable by the user 
```

**View Stock**
_No changes_

...OUT OF SCOPE - POSSIBLE FUTURE REQUIREMENTS...
```
- [ ] Sell Price should be prepopulated based on `is_default_price_list` once a Item is selected in the New Stock Modal
![image](https://github.com/user-attachments/assets/a5ba6642-cca7-4699-a647-5799c6168471)
- [ ] Cost prices are assumed to be discounted if discount would apply to this store
- [ ] Sell Prices are not editable by the user 
```

**Outbound Shipment**
- [x] Sell price should looked up from the `is_default_price_list`
- [x] However, a __discount__ should be applied based on the Customer's(Name) discount master list
   - [x] The global discount should apply from the masterlist to all items that are in that discount master list
   - [x] If an item is not in the discount master list no discount should be applied
- [x] Add Item (Frontend) => Modal runs graphql query to get Discount & apply to `invoice_line`)
- [x] Actions-> Allocate Lines (Backend: `allocatedOutboundShipmentUnallocatedLines`)  => Backend to look up default price list and apply price + any discount automatically
- [x] Add from Master List (Backend:`addToOutboundShipmentFromMasterList` ) => _No Changes required as pricing will be applied at allocation time_

**Inbound Shipment**
_No changes_

...OUT OF SCOPE -  POSSIBLE FUTURE REQUIREMENTS...
```
- [ ] Sell price should looked up from the `is_default_price_list`
- [ ] Discount will be applied based on the store name's discount list whether it comes via a transfer or otherwise
- [ ] Add from Master List (`addToInboundShipmentFromMasterList`) => look up `is_default_price_list` price for that item and apply it to `sell_price` 
- [ ] Cost price should come from sell price on inbound shipment
- [ ] Add Item  Frontend) => Modal runs graphql query to get`is_default_price_list` and applies it to `Sell Price`
- [ ] Cost prices are assumed to be discounted if discount would apply to this store
- [ ] Sell Prices are not editable by the user 
```

**Internal Orders**
_No changes required_

**Requsitions, R&R**
_No changes required_

**Returns**
_No changes required_

**Transfer aka in Processors code**
_No changes required_

**Dispensing**
- [x] Sell price should looked up from the `is_default_price_list` (Stockline Price is ignored)
- [x] No discounts are applied during dispensing

### Additional context 💌

Related to https://github.com/msupply-foundation/msupply/issues/15108

### Moneyworks Jobcode 🧰

`COMSCUST`

### PR DESCRIPTION
**Pull request by jmbrunskill**
Fixes #4703 (Part 1)

# 👩🏻‍💻 What does this PR do?

Adds two new fields to `master_list`

- [ ] is_default_price_list
- [ ] discount

Adds translation tests for new fields (using the same names are in mSupply when it is updated)
Updates test cases to use default values for master list fields.

## 💌 Any notes for the reviewer?

Using these new fields will be in the next PR... 

# 🧪 Testing

- [ ] Update mSupply to include a PR for this ISSUE -> https://github.com/msupply-foundation/msupply/issues/15108 (doesn't exist yet)
- [ ] Add a master list with is_default_price_list set
- [ ] Add a masterlist with a discount set
- [ ] Sync open-mSupply
- [ ] Check the fields are set correctly....
- [ ] Check legacy mSupply master lists still sync correctly as well

# 📃 Documentation
- [X] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour

### CODE REVIEW

File: server/repository/src/db_diesel/master_list_row.rs [hunk 1]
```diff
@@ -14,17 +14,21 @@ table! {
         code -> Text,
         description -> Text,
         is_active -> Bool,
+        is_default_price_list -> Bool,
+        discount -> Nullable<Double>,
     }
 }
 
-#[derive(Clone, Insertable, Queryable, Debug, PartialEq, Eq, AsChangeset, Default)]
+#[derive(Clone, Insertable, Queryable, Debug, PartialEq, AsChangeset, Default)]
```
Comment (jmbrunskill) [inline server/repository/src/db_diesel/master_list_row.rs:22]: `Eq` didn't like f64

Doesn't seem to have hurt anything by removing this...

File: server/repository/src/migrations/v2_03_00/mod.rs [hunk 2]
```diff
@@ -17,12 +17,12 @@ impl Migration for V2_03_00 {
...
 #[actix_rt::test]
 async fn migration_2_03_00() {
-    use v2_02_01::V2_02_01;
+    use v2_02_02::V2_02_02;
 
     use crate::migrations::*;
     use crate::test_db::*;
 
-    let previous_version = V2_02_01.version();
+    let previous_version = V2_02_02.version();
```
Comment (jmbrunskill) [inline server/repository/src/migrations/v2_03_00/mod.rs:25]: Not sure why 2_03_00 is in main but it is...

File: server/service/src/stock_line/tests/query.rs [hunk 3]
```diff
@@ -22,14 +22,14 @@ mod query {
             service.get_stock_lines(
                 &context,
                 Some(PaginationOption {
-                    limit: Some(2000),
+                    limit: Some(6000),
```
Comment (jmbrunskill) [inline server/service/src/stock_line/tests/query.rs:25]: Side quest, this test was broken in main...

**Comment by Chris-Petty** [comment]
Having a geezz cause I'll be doing the OG half of this :D but @andreievg might wanna look too.

File: server/repository/src/migrations/v2_03_00/mod.rs (hunk 2 above)
Comment (Chris-Petty) [reply server/repository/src/migrations/v2_03_00/mod.rs:25]: Yea odd! I might be mistaken, but doesn't that mean some sites might be on 2_03 as far as migrations are concerned, or is it checked against the package.json version somewhere? If not won't your new migrations be missed who are already on 2.2 (which is secretly 2_3 migration) ?

**Review by Chris-Petty — CHANGES_REQUESTED** [review]
Looks mint! Only gotcha I can think of is can you increment the inner syncv5 version? Similar (identically?) to:

https://github.com/msupply-foundation/open-msupply/pull/4716/files#diff-ed242ff60869d5532f1c4b2b3e8126822f87afe44f13b1703b67ee8989ae2834R4

Speaking of, we're gonna create conflicts with vaccine card work as we merge `main` into `develop` but they shouldn't be too drastic - just another boolean field `is_immunisation` added downstream.

File: server/repository/src/migrations/v2_03_00/mod.rs (hunk 2 above)
Comment (jmbrunskill) [reply server/repository/src/migrations/v2_03_00/mod.rs:25]: It's commented out from the migrations list, so pretty sure no one has run it...

File: package.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "open-msupply",
   "//": "Main version for the app, should be in semantic version format (any release candidate or test build should be separated by '-' i.e. 1.1.1-rc1 or 1.1.1-test",
-  "version": "2.2.01",
+  "version": "2.2.02",
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply package.json:4]: Just cargo culting here...

**Comment by jmbrunskill** [comment]
>Looks mint! Only gotcha I can think of is can you increment the inner syncv5 version? Similar (identically?) to:

Is this necessary when it's backwards compatible? Tried to read the README and KDD and it's not entirely clear.

`Central server records max and min version it's compatible with, and a simple comparison determines compatibility (`min_version <= site_version <= max_version`).`

I read that if we bump this version, OMS is no longer compatible with the current mSupply?
Where as I would have thought it would just mean we can't use this feature?

Happy to do it, as I guess it's coming in 2.3 anyway...

**Comment by Chris-Petty** [comment]
> > Looks mint! Only gotcha I can think of is can you increment the inner syncv5 version? Similar (identically?) to:
> 
> Is this necessary when it's backwards compatible? Tried to read the README and KDD and it's not entirely clear.
> 
> `Central server records max and min version it's compatible with, and a simple comparison determines compatibility (`min_version <= site_version <= max_version`).`
> 
> I read that if we bump this version, OMS is no longer compatible with the current mSupply? Where as I would have thought it would just mean we can't use this feature?
> 
> Happy to do it, as I guess it's coming in 2.3 anyway...

Hmm no that's a good point and made me think it just depends on how you write the translation - you've made them optional in the legacy record and give default values so that works without bumping the version, nice!

That said, possibly a weird pickle:

1. Have a remote site on OMS 2.0.0 (anything less than this feature) and fully synced
2. Have a Legacy Central Server on v7.17.XX (including these changes) that has `is_default_price_list = True`
3. Thus remote site hasn't integrated the new fields as OMS db doesn't have them yet
4. Upgrade remote site to version including this feature, only has default value from migration.

Thinking through this I realise the API version doesn't even help here! Nonetheless, you'd have to reintegrate the `master_list` record from the change log to populate the fields if the OMS remote is trailing behind right?

**Comment by jmbrunskill** [comment]
> > > Looks mint! Only gotcha I can think of is can you increment the inner syncv5 version? Similar (identically?) to:
> > 
> > 
> > Is this necessary when it's backwards compatible? Tried to read the README and KDD and it's not entirely clear.
> > `Central server records max and min version it's compatible with, and a simple comparison determines compatibility (`min_version <= site_version <= max_version`).`
> > I read that if we bump this version, OMS is no longer compatible with the current mSupply? Where as I would have thought it would just mean we can't use this feature?
> > Happy to do it, as I guess it's coming in 2.3 anyway...
> 
> Hmm no that's a good point and made me think it just depends on how you write the translation - you've made them optional in the legacy record and give default values so that works without bumping the version, nice!
> 
> That said, possibly a weird pickle:
> 
> 1. Have a remote site on OMS 2.0.0 (anything less than this feature) and fully synced
> 2. Have a Legacy Central Server on v7.17.XX (including these changes) that has `is_default_price_list = True`
> 3. Thus remote site hasn't integrated the new fields as OMS db doesn't have them yet
> 4. Upgrade remote site to version including this feature, only has default value from migration.
> 
> Thinking through this I realise the API version doesn't even help here! Nonetheless, you'd have to reintegrate the `master_list` record from the change log to populate the fields if the OMS remote is trailing behind right?

Yeah, I guess that is an issue. I think we just fix it by re-translating everything master_list in this update?

**Comment by andreievg** [comment]
@jmbrunskill 

> Yeah, I guess that is an issue. I think we just fix it by re-translating everything master_list in this update?

Below but for list master and list master line (for the price)

https://github.com/msupply-foundation/open-msupply/blob/385eabfd93a7728246e3aa846af5f1dede3f4873/server/repository/src/migrations/v2_02_00/item_ven.rs#L35

File: server/service/src/stock_line/tests/query.rs (hunk 3 above)
Comment (andreievg) [reply server/service/src/stock_line/tests/query.rs:25]: thanks

**Review by andreievg — CHANGES_REQUESTED** [review]
Please re-base to a new release branch v2.2.2

**Comment by jmbrunskill** [comment]
> ` [78cb8bb](/msupply-foundation/open-msupply/pull/4734/commits/78cb8bb2bc71fe727da247510da328318fdc2395)`



> @jmbrunskill
> 
> > Yeah, I guess that is an issue. I think we just fix it by re-translating everything master_list in this update?
> 
> Below but for list master and list master line (for the price)
> 
> https://github.com/msupply-foundation/open-msupply/blob/385eabfd93a7728246e3aa846af5f1dede3f4873/server/repository/src/migrations/v2_02_00/item_ven.rs#L35

Done here! (and will for next PR too)
[78cb8bb](https://github.com/msupply-foundation/open-msupply/pull/4734/commits/78cb8bb2bc71fe727da247510da328318fdc2395)

**Review by Chris-Petty — APPROVED** [review]
Haven't tested but looks good to me!

**Review by andreievg — APPROVED** [review]

**Pull request merged**